<a href="https://colab.research.google.com/github/th0void/KurtNab-Assessment/blob/main/payslip.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Payslip

A payslip is a document that shows their total wages earned for a set period.
This might be from a salary, hourly wages or commission. Payslips are also know as
‘pay advice’.  Traditionally the payslip was a paper document attached. Today
most organisations provide them electronically.

## Task 0 - Import into GitHub
First, import the notebook into Colab and save an initial copy to GitHub.

## Task 1 - Start Simple

Writing complete robust programs can take time.  Often the more we understand a
problem the easier it is to implement the solution.   One approach is to
implement the simplest version of a program and then slowly add functionality.

The aim of this notebook is to start simple and slowly build up the program.
The approach we take below is one way to understand a problem, step 1 of our
development process. For complex projects, they often follow a similar idea
where they implement a minimal program, and each week or month have intense
periods were the team adds functionality.  These advanced methodologies are
beyond the scope of an introductory programming course but start small and
slowly add functionality we do capture the essence of the process.


### The Initial (minimal) Program

Consider 'Ram' who has a salary of \$25000.  They have sold goods worth $20000
and earns 2% commission on the sales. They fall into the 10% tax bracket. We
will talk about tax-brackets in another notebook. Before making payments the
organisation needs to deduct 10% tax.

We need to program to print out a payslip for sales people.  Our payslip
is very simple, we need a title, the amount of salary, the amount of commission,
the amount of tax, and the total pay.

Run the following program for one person.

In [1]:
salary = 25000
sales = 20000
commission = 0.02 * sales
tax = (salary + commission) * 0.10
pay = salary + commission - tax
print('Payslip of Ram')
print('Salary', salary,'Commission',commission,'Tax', tax)
print('Total pay',pay)

Payslip of Ram
Salary 25000 Commission 400.0 Tax 2540.0
Total pay 22860.0


Now, consider 'Radha' who has a salary of \$30000.  They have sold goods worth $40000 and earns 2.5% commision on the sales. They fall into the 10% tax bracket and so before making payments we need to deduct 10% form his total payout.

In [2]:
salary = 30000
sales = 40000
commision = 0.025 * sales
tax = (salary + commission) * 0.10
pay = salary + commission - tax
print('Payslip of Radha')
print('Salary', salary, 'Commission', commission, 'Tax', tax)
print('Total pay', pay)

Payslip of Radha
Salary 30000 Commission 400.0 Tax 3040.0
Total pay 27360.0


## Task 2 - Generalise

What did we change from Ram to Radha?

Salary: For Ram it was $25000, while for Radha it is $30000.
Sales: For Ram it was $20000, while Radha's sales amount is $40000.
Commission Rate: Ram earns 2% commission while Radha earns 2.5% commission on sales.
The overall calculations for commission,tax and pay are adjusted and Radha's salary is $27360 while Ram's total salary is $22860.

Make what we changed as inputs (parameters) to a function

In [9]:
def generate_payslip(name, salary, sales, commission_rate):
  commission = commission_rate * sales
  tax = (salary + commission) * 0.10
  total_pay = salary + commission - tax
  print('Payslip of', name)
  print('Salary:', salary)
  print('Commission:', commission)
  print('Tax:', tax)
  print('Total pay:', total_pay)

  #Ram's information
  name_ram = 'Ram'
  salary_ram = 25000
  sales_ram = 20000
  commission_rate_ram = 0.02
  generate_payslip(name_ram, salary_ram, commission_rate_ram)

  #Radha's information
  name_radha = 'Radha'
  salary_radha = 30000
  sales_radha = 40000
  commission_rate_radha = 0.025
  generate_payslip(name_radha, salary_radha, sales_radha, commission_rate_radha)

